### PR TITLE
Move JMX statistics to an ehcache service

### PR DESCRIPTION
**Pull request by henri-tremblay**
This is the first step of a series of refactoring. It moves the statistics into ehcache.

### CODE REVIEW

File: impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+  @Override
+  public void stop() {
+    cacheManager.deregisterListener(this);
+    cacheStatistics.clear();
+  }
+
+  @Override
+  public void stateTransition(Status from, Status to) {
+    System.out.println("from " + from + " to " + to);
```
Comment (GaryWKeim) [inline impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java]: Use Logger here?

File: impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+public class DefaultStatisticsService implements StatisticsService, CacheManagerListener {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultStatisticsService.class);
+
+  private final ConcurrentMap<String, CacheStatistics> cacheStatistics = new ConcurrentHashMap<String, CacheStatistics>();
+  private volatile InternalCacheManager cacheManager;
+
+  private CacheStatistics getStatistics(String cacheName) {
+    CacheStatistics stats = cacheStatistics.get(cacheName);
+    if(stats == null) {
```
Comment (GaryWKeim) [inline impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java]: format

File: impl/src/main/java/org/ehcache/impl/internal/statistics/CacheStatistics.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+      .build();
+
+    Set<TreeNode> statResult = statQuery.execute(Collections.singleton(ContextManager.nodeFor(cache)));
+
+    if(statResult.size() < 1) {
+      throw new RuntimeException("Failed to find lowest tier statistic: " + statName + " , valid result Set sizes must 1 or more.  Found result Set size of: " + statResult.size());
+    }
+
+    //if only 1 store then you don't need to find the lowest tier
+    if(statResult.size() == 1) {
```
Comment (GaryWKeim) [inline impl/src/main/java/org/ehcache/impl/internal/statistics/CacheStatistics.java]: formatting from here to the end of this method

File: impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java (hunk 1 above)
Comment (henri-tremblay) [inline impl/src/main/java/org/ehcache/impl/internal/statistics/DefaultStatisticsService.java]: In fact, I will just remove the line. A debug line that I forgot :-(

File: impl/src/main/java/org/ehcache/impl/internal/statistics/CacheStatistics.java (hunk 3 above)
Comment (henri-tremblay) [inline impl/src/main/java/org/ehcache/impl/internal/statistics/CacheStatistics.java]: Yep. Funny, some of these are cut & paste

**Comment by henri-tremblay** [comment]
I don't know what is our politic right now. But we still have a lot of author tags everywhere. Depending on the project, people do or do not use them.

**Pull request merged**